src/functions/casing.ts

```diff
@@ -1,40 +1,95 @@
 import { Dictionary } from '@stoplight/types';
-import { ICasingOptions, IFunction } from '../types';
-
-const CASES: Dictionary<RegExp, ICasingOptions['type']> = {
-  flat: /^[a-z]+$/,
-  camel: /^[a-z]+(?:[A-Z][a-z]+)*$/,
-  pascal: /^[A-Z][a-z]+(?:[A-Z][a-z]+)*$/,
-  kebab: /^[a-z]+(?:-[a-z]+)*$/,
-  cobol: /^[A-Z]+(?:-[A-Z]+)*$/,
-  snake: /^[a-z]+(?:_[a-z]+)*$/,
-  macro: /^[A-Z]+(?:_[A-Z]+)*$/,
-};
+import { IFunction, IRule, RuleFunction } from '../types';
+
+export enum CasingType {
+  flat = 'flat',
+  camel = 'camel',
+  pascal = 'pascal',
+  kebab = 'kebab',
+  cobol = 'cobol',
+  snake = 'snake',
+  macro = 'macro',
+}
+
+interface ICasingOptionsOverrides {
+  disallowDigits?: boolean;
+  separator?: string;
+  allowLeadingSeparator?: boolean;
+}
+
+export interface ICasingOptions extends ICasingOptionsOverrides {
+  type: CasingType;
+}
 
-const CASES_WITH_DIGITS: Dictionary<RegExp, ICasingOptions['type']> = {
-  flat: /^[a-z][a-z0-9]*$/,
-  camel: /^[a-z][a-z0-9]*(?:[A-Z0-9][a-z0-9]+)*$/,
-  pascal: /^[A-Z][a-z0-9]*(?:[A-Z0-9][a-z0-9]+)*$/,
-  kebab: /^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$/,
-  cobol: /^[A-Z][A-Z0-9]*(?:-[A-Z0-9]+)*$/,
-  snake: /^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$/,
-  macro: /^[A-Z][A-Z0-9]*(?:_[A-Z0-9]+)*$/,
+export type CasingRule = IRule<RuleFunction.CASING, ICasingOptions>;
+
+const CASES: Dictionary<string, CasingType> = {
+  [CasingType.flat]: '[a-z][a-z{__DIGITS__}]*',
+  [CasingType.camel]: '[a-z][a-z{__DIGITS__}]*(?:[A-Z{__DIGITS__}][a-z{__DIGITS__}]+)*',
+  [CasingType.pascal]: '[A-Z][a-z{__DIGITS__}]*(?:[A-Z{__DIGITS__}][a-z{__DIGITS__}]+)*',
+  [CasingType.kebab]: '[a-z][a-z{__DIGITS__}]*(?:-[a-z{__DIGITS__}]+)*',
+  [CasingType.cobol]: '[A-Z][A-Z{__DIGITS__}]*(?:-[A-Z{__DIGITS__}]+)*',
+  [CasingType.snake]: '[a-z][a-z{__DIGITS__}]*(?:_[a-z{__DIGITS__}]+)*',
+  [CasingType.macro]: '[A-Z][A-Z{__DIGITS__}]*(?:_[A-Z{__DIGITS__}]+)*',
 };
 
 export const casing: IFunction<ICasingOptions> = (targetVal, opts) => {
   if (typeof targetVal !== 'string' || targetVal.length === 0) {
     return;
   }
 
-  const set = opts.disallowDigits ? CASES : CASES_WITH_DIGITS;
+  if (!validate(opts)) {
+    return;
+  }
+
+  const casingValidator = buildFrom(CASES[opts.type], opts);
+
+  if (casingValidator.test(targetVal)) {
+    return;
+  }
+
+  return [
+    {
+      message: `must be ${opts.type} case`,
+    },
+  ];
+};
+
+const validate = (opts: ICasingOptions): boolean => {
+  if (!(opts.type in CASES)) {
+    // TODO: throw a relevant error message
```

Comment:
@philsturgeon I'd really like to throw here. If the type is unknown, there are great chances it's because of a ruleset design mistake. And rather than silently ignoring it, I'd rather let the user know about it.

Reply:
You can throw, no harm will be done.

Reply:
Fixed